### ISSUE TITLE
Calling String_val when the runtime system is released is not safe

### ISSUE THREAD

**Issue by yallop** [andrenth/ocaml-milter#2]
[This code](https://github.com/andrenth/ocaml-milter/blob/e90c73ead212aef28be6b8fd417c9487bb6789b6/lib/milter_stubs.c#L673):

```C
    caml_release_runtime_system();
    ret = smfi_setconn(String_val(conn_value));
    caml_acquire_runtime_system();
```

should copy `conn_val` (using `strdup` or otherwise) before releasing the runtime system, since `String_val` accesses the OCaml heap.

**Comment by yallop** [comment]
There are some other similar issues, e.g. [here](https://github.com/andrenth/ocaml-milter/blob/e90c73ead212aef28be6b8fd417c9487bb6789b6/lib/milter_stubs.c#L1114-L1116):

```C
    char *macros = String_val(macros_val);

    caml_release_runtime_system();
    ret = smfi_setsymlist(ctx, stage, macros);
    caml_acquire_runtime_system();
```
where `macros` is pointing into the OCaml heap and (presumably) accessed by `smfi_setsymlist` while the runtime lock is not held.

**Comment by andrenth** [comment]
Thanks. I believe I've fixed all cases in 1432e10.
